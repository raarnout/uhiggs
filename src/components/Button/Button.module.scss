.btn-square {
    border-radius: 0;
}

.btn-pill {
    border-radius: 10rem;
}

.btn-shadowed {
    box-shadow: 4px 4px 8px rgba(0, 0, 0, 0.15);
    transition: box-shadow 0.2s ease, transform 0.2s ease;
  
    &:active {
      box-shadow: 1px 1px 4px rgba(0, 0, 0, 0.2);
      transform: translateY(1px);
    }
}     