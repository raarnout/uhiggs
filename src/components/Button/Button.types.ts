import React from 'react';

export type BootstrapVariant =
  | 'primary'
  | 'secondary'
  | 'success'
  | 'danger'
  | 'warning'
  | 'info'
  | 'light'
  | 'dark'
  | 'link';

export type BootstrapSize = 'sm' | 'md' | 'lg';

export type BootstrapOutline = true | false;

export type BootstrapDisabled = true | false;

export type BootstrapBlock = true | false;

export type Square = true | false;

export type Pill = true | false;

export type Shadow = true | false;

export interface ButtonProps
  extends React.ButtonHTMLAttributes<HTMLButtonElement> {
  variant?: BootstrapVariant;
  size?: BootstrapSize;
  outline?: BootstrapOutline;
  disabled?: BootstrapDisabled;
  block?: BootstrapBlock;
  square?: Square;
  pill?: Pill;
  shadow?: Shadow;
  iconStart?: string;
  iconEnd?: string;
  className?: string;
}
