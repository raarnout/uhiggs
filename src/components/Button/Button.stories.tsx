import type { Meta, StoryObj } from '@storybook/react';
import { Button } from './Button';

const meta: Meta<typeof Button> = {
  title: 'Components/Button',
  component: Button,
  tags: ['autodocs'],
  argTypes: {
    onClick: { action: 'clicked' },
    variant: {
      control: { type: 'radio' },
      options: ['primary',
        'secondary',
        'success',
        'danger',
        'warning',
        'info',
        'light',
        'dark',
        'link'
      ]
    },
    size: {
      control: { type: 'radio' },
      defaultValue: 'md',
      options: ['sm', 'md', 'lg'], 
      labels: {
        sm: 'Small',
        md: 'Medium',
        lg: 'Large'
      }
    },
    outline: {
      control: { type: 'boolean' },
      defaultValue: false,
    },
    square: {
      control: { type: 'boolean' },
      defaultValue: false,
    },
    pill: {
      control: { type: 'boolean' },
      defaultValue: false,
    },
    shadow: {
      control: { type: 'boolean' },
      defaultValue: false,
    },
    disabled: {
      control: { type: 'boolean' },
      defaultValue: false,
    },
    block: {
      control: { type: 'boolean' },
      defaultValue: false,
    }
  },
  parameters: {
    docs: {
      description: {
        story: 'Deze knop toont een icoon aan het begin Ã©n aan het einde van de knoptekst.',
      },
    },
  },
};

export default meta;
type Story = StoryObj<typeof Button>;

export const button: Story = {
  args: {
    children: 'Button',
    variant: 'primary',
    size: 'md',
    outline: false,
    disabled: false,
    square: false,
    pill: false,
    shadow: false,
    block: false,
    iconStart: 'cloud-download',
    iconEnd: 'arrow-right',
  },
};

