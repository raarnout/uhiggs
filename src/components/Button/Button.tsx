import React from 'react';
import classNames from 'classnames';
import { ButtonProps } from './Button.types';
import styles from './Button.module.scss'; 

export const Button: React.FC<ButtonProps> = ({
  children,
  onClick,
  variant = 'primary',
  size,
  outline = false,
  disabled = false,
  block = false,
  square = false,
  pill = false,
  shadow = false,
  className,
  ...props
}) => {
  const btnClass = classNames(
    'btn',
    outline ? `btn-outline-${variant}` : `btn-${variant}`,
    size ? `btn-${size}` : '',
    square ? styles['btn-square'] : '',
    pill ? styles['btn-pill'] : '',
    shadow ? styles['btn-shadowed'] : '', 
    block ? `w-100 d-block` : '',
    className
  );

  return (
    <button 
      className={btnClass} 
      onClick={onClick} 
      disabled={disabled}
      {...props}
    >
      {children}
    </button>
  );
};
