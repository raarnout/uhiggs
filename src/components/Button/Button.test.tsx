import { render, screen, fireEvent } from '@testing-library/react';
import { Button } from './Button';
import { axe } from 'jest-axe'; 
import userEvent from '@testing-library/user-event'; 

describe('Button component', () => {
  test('renders children', () => {
    render(<Button>Click me</Button>);
    expect(screen.getByText('Click me')).toBeInTheDocument();
  });

  test('calls onClick handler', () => {
    const onClick = jest.fn();
    render(<Button onClick={onClick}>Click</Button>);
    fireEvent.click(screen.getByText('Click'));
    expect(onClick).toHaveBeenCalled();
  });

  test('applies variant class', () => {
    render(<Button variant="danger">Delete</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-danger');
  });

  test('applies outline class when outline=true', () => {
    render(<Button variant="success" outline={true}>Save</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-outline-success');
  });

  test('applies size class', () => {
    render(<Button size="lg">Large</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-lg');
  });

  test('applies square class', () => {
    render(<Button square>Square</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-square');
  });

  test('applies shadow class', () => {
    render(<Button shadow>Shadow</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-shadowed');
  });

  test('applies pill class', () => {
    render(<Button pill>Pill</Button>);
    expect(screen.getByRole('button')).toHaveClass('btn-pill');
  });

  test('renders iconStart', () => {
    render(<Button iconStart="check">OK</Button>);
    expect(document.querySelector('.bi.bi-check')).toBeInTheDocument();
  });

  test('renders iconEnd', () => {
    render(<Button iconEnd="arrow-right">Next</Button>);
    expect(document.querySelector('.bi.bi-arrow-right')).toBeInTheDocument();
  });

  test('adds block class', () => {
    render(<Button block>Full width</Button>);
    expect(screen.getByRole('button')).toHaveClass('w-100');
  });

  test('is disabled when disabled is true', () => {
    render(<Button disabled>Disabled</Button>);
    expect(screen.getByRole('button')).toBeDisabled();
  });

  test('applies custom className', () => {
    render(<Button className="custom-class">Styled</Button>);
    expect(screen.getByRole('button')).toHaveClass('custom-class');
  });
});

describe('Button accessibility', () => {
    test('has no a11y violations (default)', async () => {
      const { container } = render(<Button>Click me</Button>);
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  
    test('has no a11y violations when disabled', async () => {
      const { container } = render(<Button disabled>Disabled</Button>);
      expect(screen.getByRole('button')).toBeDisabled();
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  
    test('has no a11y violations when using iconStart and iconEnd', async () => {
      const { container } = render(
        <Button iconStart="check" iconEnd="arrow-right">Submit</Button>
      );
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  
    test('has no a11y violations for icon-only button with aria-label', async () => {
      const { container } = render(
        <Button iconStart="check" aria-label="Confirm action" />
      );
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  
    test('fails a11y test if icon-only button has no label', async () => {
      const { container } = render(<Button iconStart="check" />);
      const results = await axe(container);
      expect(results.violations.length).toBeGreaterThan(0);
    });
  
    test('has accessible role and is focusable', () => {
      render(<Button>Focusable</Button>);
      const button = screen.getByRole('button');
      expect(button).toBeInTheDocument();
    });

    test('is focusable via keyboard (Tab)', async () => {
    render(
        <>
        <button>Before</button>
        <Button>Focusable</Button>
        <button>After</button>
        </>
    );

    const user = userEvent.setup();
    await user.tab(); // focus op "Before"
    await user.tab(); // focus op Button

    expect(screen.getByRole('button', { name: 'Focusable' })).toHaveFocus();
    });
});

describe('Button snapshot', ()=> {
    test('matches snapshot', () => {
        const { container } = render(
          <Button
            variant="primary"
            size="lg"
            shadow={true}
            square={true}   
            iconStart="check"
            iconEnd="arrow-right"
          >
            Snapshot Button
          </Button>
        );
      
        expect(container).toMatchSnapshot();
      });
})


