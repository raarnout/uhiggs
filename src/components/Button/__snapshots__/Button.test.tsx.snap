// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button snapshot matches snapshot 1`] = `
<div>
  <button
    class="btn btn-primary btn-lg btn-square btn-shadowed btn-icon-wrapper"
  >
    <i
      aria-hidden="true"
      class="bi bi-check btn-icon"
    />
    Snapshot Button
    <i
      aria-hidden="true"
      class="bi bi-arrow-right btn-icon"
    />
  </button>
</div>
`;
